polygon/sync/service.go

```diff
@@ -52,7 +55,13 @@ func NewService(
 	blocksVerifier := VerifyBlocks
 	p2pService := p2p.NewService(logger, maxPeers, sentryClient, statusDataProvider.GetStatusData)
 	execution := newExecutionClient(executionClient)
-	store := NewStore(logger, execution, bridgeService)
+
+	signaturesCache, err := lru.NewARC[common.Hash, common.Address](InMemorySignatures)
+	if err != nil {
+		panic(err)
+	}
+
+	store := NewStore(logger, execution, bridgeService, NewWiggleCalculator(borConfig, signaturesCache, heimdallService))
```

Comment:
To me it doesn't make sense for the data store to have this metric calculator - I think it should be in `Connect` in `CanonicalChainBuilder`.

Reply:
this was a suggestion from Milen in his task requirements. https://github.com/erigontech/erigon/issues/12308 (option 2)

Reply:
That issue has three options - I think option 3 is the cleanest because you will only have one thread using `signaturesCache`. With this solution you may have a race condition between the store and sync goroutines. 

Option 1 from the issue description is fine by me too since it doesn't introduce this into the store, however I think Option 3 is best since we would have just one user of signaturesCache.

Reply:
I agree with Milen that it is better to keep the chain builder as slim as it is possible. signatureCache is a thread safe structure in fact, it has lock/unlock mechanism inside to protect its data from parallel access, that s why there is no risk of race conditions. 

So, what is our solution here? Because I am still leaning to not move it to the chainBuilder code because of the arguments in the task description.

Reply:
> it has lock/unlock mechanism inside to protect its data from parallel access

What is the performance impact of this?

Reply:
Discussed offline and came to a conclusion that it's best to place it in `applyNewBlockOnTip` for now and test any performance degradation coming from the `RWMutex` in `ARCCache`.